File: swift/easy/234-palindrome-linked-list.swift
func isPalindrome(_ head: ListNode?) -> Bool {
    if (head == nil) {
        return true
    }

    let middle = findMiddle(head)
    var last = reverse(middle)
    var first = head

    while (first != nil && last != nil) {
        if (first!.val != last!.val) {
            return false
        } else {
            first = first!.next
            last = last!.next
        }
    }

    return true
}

func findMiddle(_ head: ListNode?) -> ListNode? {
    var slow = head
    var fast = head

    while (fast?.next != nil) {
        slow = slow!.next
        fast = fast!.next!.next
    }

    return slow
}

func reverse(_ head: ListNode?) -> ListNode? {
    var prev: ListNode? = nil
    var current = head

    while (current != nil) {
        let next = current!.next
        current!.next = prev
        prev = current
        current = next
    }

    return prev
}

File: swift/easy/234.palindrome-linked-list.java
public boolean isPalindrome(ListNode head) {
    ListNode first = head;

    ListNode middle = findMiddleNode(head);
    ListNode second = reverse(middle);

    while (first != null && second != null) {
        if (first.val != second.val) {
            return false;
        }

        first = first.next;
        second = second.next;
    }

    return true;
}

private ListNode findMiddleNode(ListNode head) {
    ListNode tail = head;

    while (tail != null && tail.next != null) {
        head = head.next;
        tail = tail.next.next;
    }

    return head;
}

private ListNode reverse(ListNode head) {
    ListNode prev = null;

    while (head != null) {
        ListNode next = head.next;
        head.next = prev;
        prev = head;
        head = next;
    }

    return prev;
}
